server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -1142,5 +1257,64 @@ synchronized void markItemAsDropped(int slot) {
             );
             itemResponses.add(BulkItemResponse.success(slot, docWriteRequest.opType(), dropped));
         }
+
+        /**
+         * Mark the document at the given slot in the bulk request as having failed in the ingest service. The document will be redirected
+         * to a data stream's failure store.
+         * @param slot the slot in the bulk request to redirect.
+         * @param targetIndexName the index that the document was targeting at the time of failure.
+         * @param e the failure encountered.
+         */
+        public void markItemForFailureStore(int slot, String targetIndexName, Exception e) {
+            if (DataStream.isFailureStoreEnabled() == false) {
+                markItemAsFailed(slot, e);
```

Comment:
Given the definition of `shouldStoreFailure` it seems like it would be wrong to ever get through this part of the `if` -- should this be asserted against at dev time?